generator/src/swam/generator/ImportGenerator.scala

```diff
@@ -0,0 +1,163 @@
+package swam
+package generator
+
+import java.nio.file.{Paths, StandardOpenOption}
+
+import swam.ValType.{F32, F64, I32, I64}
+import swam.runtime.Import
+
+import scala.reflect.runtime.universe._
+import org.scalafmt.interfaces.Scalafmt
+import org.fusesource.scalate.TemplateEngine
+import cats.effect.{Blocker, Effect, IO}
+import cats.implicits._
+import cats.effect._
+import fs2.{text, _}
+
+/**
+  * @author Javier Cabrera-Arteaga on 2020-03-06
+  * Generates the scala code with the template to use as Import in the WASM execution
+  */
+class ImportGenerator[F[_]: Effect](implicit cs: ContextShift[F]) {
+
+  val f = "IO"
+  val scalafmt = Scalafmt.create(this.getClass.getClassLoader)
+  val config = Paths.get(".scalafmt.conf")
+
+  private var templateFile = getClass.getClassLoader.getResource("import_template.mustache").getFile
+  var registeredFunctions: Set[String] = Set[String]()
+
+  /**
+    * Map WASM to scala primitive types
+    * @param v
+    * @return
+    */
+  private def getScalaType(v: ValType): String = v match {
+    case I32 => s"Int"
+    case I64 => s"Long"
+    case F32 => s"Float"
+    case F64 => s"Double"
+  }
+
+  /**
+    * Change the template file
+    * @param file
+    */
+  def changeTemplate(file: String) = templateFile = file
+
+  /**
+    * Creates the string representing the scala function return based on the arity of the return object in WASM
+    * @param t
+    * @return
+    */
+  private def buildReturn(t: Vector[ValType]): String = {
+    if (t.isEmpty)
+      "Unit"
+    else {
+      if (t.length == 1)
+        getScalaType(t.last)
+      else
+        t.map(getScalaType).mkString("(", ",", ")")
+    }
+  }
+
+  /**
+    * Generates a context Map to feed the template rendering
+    * @param imports
+    * @return
+    */
+  def getContext(imports: Vector[Import]): Seq[Map[String, Any]] = {
+    val sorted = imports
+      .collect { case i: Import.Function => i } // Filter by function type
+      .groupBy(t => t.moduleName) // Group by module
+      .view
+      .mapValues(_.toSet) // remove duplicated entries
+
+    // Generating DTO
+    val dto = sorted.keys.zipWithIndex
+      .map(
+        m =>
+          Map(
+            "module" -> m._1,
+            "comma" -> (m._2 < sorted.keys.size - 1),
+            "fields" -> sorted(m._1).toSeq.zipWithIndex
+              .map(
+                f => {
+                  Map(
+                    "name" -> f._1.fieldName,
+                    "return" -> buildReturn(f._1.tpe.t),
+                    "params" -> f._1.tpe.params.zipWithIndex.map(t => s"p${t._2}:${getScalaType(t._1)}").mkString(","),
+                    "comma" -> (f._2 < sorted(m._1).size - 1)
+                  )
+                }
+              )
+          ))
```

Comment:
When dealing with tuples, it is a bit more elegant (I think) to use the partial function syntax to deconstruct the parameters. Something like this (not compiled): 
```suggestion
      .map {
        case (name, idx) =>
          Map(
            "module" -> name,
            "comma" -> (idx < sorted.keys.size - 1),
            "fields" -> sorted(name).toSeq.zipWithIndex
              .map {
               case (Function(moduleName, fieldName, tpe), idx) =>
                  Map(
                    "name" -> fieldName,
                    "return" -> buildReturn(tpe.t),
                    "params" -> tpe.params.zipWithIndex.map(t => s"p${t._2}:${getScalaType(t._1)}").mkString(","),
                    "comma" -> (idx < sorted(name).size - 1)
                  )
              }
          )
        }
```